peer/brontide.go

```diff
@@ -3656,10 +3643,41 @@ func (p *Brontide) handleCloseMsg(msg *closeMsg) {
 		return
 	}
 
-	// Queue any messages to the remote peer that need to be sent as a part of
-	// this latest round of negotiations.
-	for _, msg := range msgs {
-		p.queueMsg(msg, nil)
+	// If the link is nil we can skip this as it means we have already torn
+	// it down.
+	if link != nil {
+		// Queue any messages to the remote peer that need to be sent as
+		// a part of this latest round of negotiations.
+		oShutdown.WhenSome(
+			// If we have a Shutdown to send, let's defer it until
+			// the next fully committed state.
+			func(msg lnwire.Shutdown) {
+				link.OnCommitOnce(htlcswitch.Outgoing, func() {
+					// Disable outgoing adds in preparation
+					// for sending our Shutdown.
+					err := link.DisableAdds(
+						htlcswitch.Outgoing,
+					)
+					if err != nil {
+						p.log.Warn("outgoing link adds"+
+							" already disabled: %v",
+							link.ChanID())
+					}
+
+					p.queueMsg(&msg, nil)
+				})
+			},
+		)
+		oClosingSigned.WhenSome(
+			// If we have a ClosingSigned to send, let's defer it
+			// until the channel has no remaining HTLCs.
+			func(msg lnwire.ClosingSigned) {
+				link.OnFlushedOnce(func() {
+					p.cfg.Switch.RemoveLink(msg.ChannelID)
+					p.queueMsg(&msg, nil)
```

Comment:
I think there are some subtle issues here because the counter-party is allowed to interact w/ our `ChanCloser` state machine by sending `closing_signed` even though we're "not ready" for it:

- If the counter-party repeatedly sends us `closing_signed` that we don't agree with, we'll respond with our own `closing_signed`. Here, respond means we add another flush hook to the link and keep our `closing_signed` response in memory. When we are finally ready to flush, we could end up sending hundreds of thousands of messages assuming the node hasn't OOM'd first.
- If the counter-party instead sends us a message that agrees with our `closing_signed` or is close enough to ours that we agree with them, we'll end up broadcasting the cooperative close transaction even though the link is still up. Due to the way balance is calculated in coop close, if we've offered an HTLC to the counter-party, that HTLC will effectively be deleted from our balance and be used as fees. This could be used as a griefing vector

Reply:
The protocol adherence of the shutdown procedure should be enforced by the ChanCloser, no?

I do agree here that if there are many rounds of closing_signed negotiation, we need to ensure that the link is only given a single flush hook to tear itself down. If the link is already torn down it means that we are firmly in closing negotiation and we can synchronously send the message back.

Nice catch.